README.md
# MorseWav
Pearl script suite to convert text files to Morse code in a .wav file
 
 See "instruction.txt" for instructions on how to embed controls into the source text file for conversion.
 While aimed at tone creation, the script also (originally) converts text to Morse code.
 


Audio-Wav-0.13/Changes
Revision history for Perl extension Audio::Wav.

0.13  Fri Mar 30 21:17:18 GMT 2012
  - read: don't call close() at DESTROY time when the file handle
    creationfailed during new() (Murali Ramanathan)
      https://rt.cpan.org/Ticket/Display.html?id=71462
  - apply patch to allow read/write of files from handle
      https://rt.cpan.org/Ticket/Display.html?id=71715
  - fix problem with $^X under mod_perl (thanks, mario and david!)
      https://rt.cpan.org/Ticket/Display.html?id=62060
  - fix bug in perldoc (Sergei Steshenko)
      https://rt.cpan.org/Ticket/Display.html?id=57114
  - fix problem with writing 24,32 bps files reported by Sergei Steshenko:
      http://rt.cpan.org/Public/Bug/Display.html?id=57093
  - spelling fixes from debian
      http://rt.cpan.org/Public/Bug/Display.html?id=69644

0.12  Sun May 30 08:45:03 GMT 2010
  - don't use values > MAXINT. +2147483648 is not a valid unsigned
    int32 (it's MAXINT+1). while we're at it, put all constants in
    the XS in hex format for clarity. possibly fixes:
      http://www.cpantesters.org/cpan/report/6945385
  - fall back to pureperl when Inline::C is installed but not functional
    as reported by Michel Lanne.

0.11  Fri Mar 12 04:12:44 GMT 2010
  - fix two bugs triggered with inline on win32 - one related to assuming 
    C99 support and one related to lack of stdint.h or equivalent. tested
    with strawberry perl, should work with msvc too. bug report here:
      http://www.cpantesters.org/cpan/report/6920668
    same smoker reports tests with 0.11 here:
      http://www.cpantesters.org/cpan/report/6930825 

0.10  Sun Feb 14 04:09:00 GMT 2010
  - add COPYRIGHT information for debian folks
  - include xt/

0.09  Thu Feb 11 14:58:37 GMT 2010
  - Support 24 and 32 bit wav reading w/o Inline::C (from Wolfram humann)
    ( see https://rt.cpan.org/Public/Bug/Display.html?id=36452 )
  - add a META.yml and some other easy kwalitee tweaks
  - Audio::Wav::Read::_has_inline moved to Audio::Wav::_has_inline so
    it can be queried before instantiating a reader (and later be used
    internally for Audio::Wav::Write)

0.08  Tue Feb 09 06:29:43 GMT 2010
  - fix regression: read() returned bogus samples when Inline::C not available

0.07  Sun Feb 07 18:05:41 GMT 2010
  - change API so you can call Audio::Wav->{read|write} w/o new() if preferred
  - increase pure perl read speed by a factor of ~2.4
    * remove unnecessary bounds check
    * put block in closure, avoiding double hash lookup
    * put $block assign in _init_read_sub, put read_sub in closure, too
    * pull $handle into closure:
    * get rid of closure, and get rid of read() - inline it with $read_sub
  - use Inline::C (if available) to increase read speed by a factor of ~2.3
  - experimental support for reading 24- and 32- bit data (suspected to
    work on little endian machines that use Inline::C)

0.06  Wed Mar 22 12:00:00 2006
  - Fixed a circular reference in Audio::Wav::Write::Header that was causing
    memory to leak (thanks Sumitro Chowdhury).
  - Tidied up bits and pieces.
  - Added very basic support for WAVEFORMATEXTENSIBLE.
  - When writing files, finish() will now be called by DESTROY if necessary.

0.05  Tue Oct 25 12:20:00 2005
  - Audio::Wav::Read::position_samples should have divided by block_align
    rather than multiplied (thanks David Brandt).
  - Fixed bug where unknown blocks weren't skipped (thanks Robert Hiller).

0.04  Thu Dec 30 07:47:00 2004
  - fixed a bug in Audio::Wav::Read::move_to, now adds where the data
    actually starts to the position given.
  - Audio::Wav::Read::move_to now rereads data length to see if file has
    grown since this was last read.
  - added method Audio::Wav::Read::reread_length, rereads the length of
    the file in case it is being written to as we are reading it.
  - added method Audio::Wav::Read::read_raw_samples which will read X
    samples in raw format.
  - added method Audio::Wav::Read::position_samples which returns the
    current audio data position in samples.
  - in method Audio::Wav::Write::add_cue, if sample position supplied
    is undefined, then the position will be the current position (end
    of all data written so far).
  - in method Audio::Wav::Write, moved the option of not caching data
    from the write_raw method to new.

0.03  Fri Jun 11 13:29:00 2004
  - minor bug fix to pass tests with Perl 5.8.3 (thanks to Jeremy Devenport).

0.02  Sat Sep 01 15:15:00 2001
  - works on big endian machines!
  - no need for Audio::Tools anymore
  - added support for info & sampler blocks.
  - now honours padding bytes
  - read & read_raw no longer return non-audio data.
  - added error handler.
  - slight speed improvement for read & write methods
  - some other fixes.

0.01  Fri Dec 11 05:54:22 1998
  - original version; created by h2xs 1.18



Audio-Wav-0.13/test.pl
use strict;

$| = 1;

my $started = time;

my $out_dir = 'test_output';

unless ( -d $out_dir ) {
    mkdir( $out_dir, 0777 ) ||
	die "unable to make test output directory '$out_dir' - ($!)";
}

use Audio::Wav;

my $cnt = 0;

print "1..4\n\n";

print "NOTE: ".($Audio::Wav::_has_inline ? 'YES' : 'NOT')." using inline\n";

### Wav Creation

print "\nTesting wav creation\n";

my %wav_options = ( # these are optional & default to 0
    '.01compatible'   => 0,
    'oldcooledithack' => 0,
    'debug'           => 0,
);


my $wav = Audio::Wav -> new( %wav_options );

my $file_out = $out_dir . '/testout.wav';
my $file_copy = $out_dir . '/testcopy.wav';
my $sample_rate = 11025;
my $bits_sample = 8;
#my $bits_sample = 32; warn "non-8-bit test";
my $length = 2;
my $channels = 1;
#my $channels = 2; warn "stereo test";

my $details =	{
    'bits_sample'	=> $bits_sample,
    'sample_rate'	=> $sample_rate,
    'channels'		=> $channels,
};

my $write = $wav -> write( $file_out, $details );

&add_slide( 50, 300, $length );

$write -> set_info( 'software' => 'Audio::Wav' );

my $marks = $length / 3;
foreach my $xpos ( 1 .. 2 ) {
    my $ypos = &seconds_to_samples( $xpos * $marks );
    $write -> add_cue( $ypos, "label $xpos", "note $xpos" );
    print "Cue $xpos at $ypos\n";
}

my $sec_samps = &seconds_to_samples( 1 );
$write -> add_cue( $sec_samps, "onesec", "one second" );
print "Cue 3 at $sec_samps\n";

my %samp_loop = (
    'start'	=> &seconds_to_samples( $length * .25 ),
    'end'	=> &seconds_to_samples( $length * .75 ),
);

$write -> add_sampler_loop( %samp_loop );

my %display = (
    'id'	=> 1,
    'data'	=> 'Submarine Captain',
);

$write -> add_display( %display );

$write -> finish();

$cnt ++;
print "ok $cnt\n";

### Wav Copying

print "\nTesting wav copying and shortcut syntax\n";

my $read = Audio::Wav -> read( $file_out );

# print Data::Dumper->Dump([ $read -> details() ]);

$write = $wav -> write( $file_copy, $read -> details() );

my $cues = $read -> get_cues();

for my $id ( 1 .. 3 ) {
    print "Cue $id at ", $cues -> {$id} -> {'position'}, "\n";
}

my $buffer = 512;
my $total = 0;
$length = $read -> length();

while ( $total < $length ) {
    my $left = $length - $total;
    $buffer = $left unless $left > $buffer;
    my $data = $read -> read_raw( $buffer );
    last unless defined( $data );
    $write -> write_raw( $data, $buffer );
    $total += $buffer;
}

$write -> finish();

$cnt ++;
print "ok $cnt\n";

### Wav Comparing

print "\nComparing wav files $file_out & $file_copy\n";

my $file_orig = $file_out;

open ORIG, $file_orig or die "Can't open file '$file_orig': $!\n";
binmode ORIG;

my $data_orig;
while (<ORIG>) {
  $data_orig.=$_;
}
close ORIG;

open COPY, $file_copy or die "Can't open file '$file_copy': $!\n";
binmode COPY;

my $data_copy;
while (<COPY>) {
  $data_copy.=$_;
}
close COPY;


if (length($data_copy) ne length($data_orig)) {
    die "Wav files ARE NOT identical; they are of different lengths";
}


if ($data_copy ne $data_orig) {
    die "Wav files ARE NOT identical";
}

$cnt ++;
print "ok $cnt\n";

print "\nTesting sample wav file\n";

if ( &test_wav() ) {
    print "sample wav file was read correctly\n";
} else {
    die "sample wav file was not read correctly\n";
}

$cnt ++;
print "ok $cnt\n";

print "took ", int( time - $started ), " seconds";

sub test_wav {
    my $file = 'test_tone.wav';
    my $cued_sample = -12;
    my %match_details = (
	'bits_sample'	=> 8,
	'length'	=> '0.5',
	'block_align'	=> 1,
	'bytes_sec'	=> 8000,
	'total_length'	=> 4152,
	'channels'	=> 1,
	'sample_rate'	=> 8000,
	'data_length'	=> 4000,
	'data_start'	=> '44',
    );
    my $read = $wav -> read( $file );
    my $details = $read -> details();
    foreach my $key ( keys %match_details ) {
	my( $want, $is ) = ( $details -> {$key}, $match_details{$key} );
	next if $details -> {$key} eq $match_details{$key};
	warn "mismatched value for $key, wanted $want, but got $is\n";
	return 0;
    }
    my $cues = $read -> get_cues();
    unless ( exists $cues -> {'1'} ) {
	warn "no cues found in $file\n";
	return 0;
    }
    my $pos = $cues -> {'1'} -> {'position'};
    unless ( $read -> move_to( $pos ) ) {
	warn "unable to move to sample $pos\n";
	return 0;
    }
    my( $sample ) = $read -> read();
    unless ( $cued_sample == $sample ) {
	warn "sample at position $pos does not match $cued_sample (should be $sample)\n";
	return 0;
    }
    return 1;
}

sub add_slide {
    my $from_hz = shift;
    my $to_hz = shift;
    my $length = shift;
    my $volume = .5;
    my $diff_hz = $to_hz - $from_hz;
    my $pi = ( 22 / 7 ) * 2;
    $length *= $sample_rate;
    my $max_no =  ( 2 ** $bits_sample ) / 2;
    my $half = int( $length / 2 );
    my $pos = 0;
    foreach my $rev ( 0, 1 ) {
	my $target = $half;
	$target *= 2 if $rev;
	while ( $pos < $target ) {
	    $pos ++;
	    my $rev_pos = $rev ? ( $half - ( $pos - $half ) ) : $pos;
	    my $prog = $rev_pos / $half;
	    my $hz = $from_hz + ( $diff_hz * $prog );
	    my $cycle = $sample_rate / $hz;
	    my $mult = $rev_pos / $cycle;
	    my $samp = sin( $pi * $mult ) * $max_no;
	    $samp *= $volume;
	    $write -> write( map $samp, 1 .. $channels );
	}
    }

}

sub seconds_to_samples {
    my $time = shift;
    return $time * $sample_rate;
}



Audio-Wav-0.13/Wav/Tools.pm
package Audio::Wav::Tools;

use strict;
eval { require warnings; }; #it's ok if we can't load warnings

use vars qw( $VERSION );
$VERSION = '0.13';

sub new {
    my ($class, %options) = @_;
    my $self = {
        'errorHandler' => undef,
    };

    foreach my $key ( qw( .01compatible oldcooledithack debug ) ) {
        $self -> {$key} = exists( $options{$key} ) && $options{$key} ? 1 : 0;
    }
    bless $self, $class;
    return $self; 
}

sub is_debug {
    my $self = shift;
    return $self -> {'debug'};
}

sub is_01compatible {
    my $self = shift;
    return $self -> {'.01compatible'};
}

sub is_oldcooledithack {
    my $self = shift;
    return $self -> {'oldcooledithack'};
}

sub set_error_handler {
    my $self = shift;
    my $handler = shift;
    unless ( ref( $handler ) eq 'CODE' ) {
        die 'set_error_handler is expecting a reference to a sub routine';
    }
    $self -> {'errorHandler'} = $handler; 
}

sub is_big_endian {
    my $self = shift;
    return $self -> {'is_big_endian'} if exists $self -> {'is_big_endian'};
    my $VALUE = 1801677134;
    my $nativeLong  = pack 'L', $VALUE;   # 'kciN' (big) or 'Nick' (little)
    my $bigLong     = pack 'N', $VALUE;   # should return 'kciN'
    $self -> {'is_big_endian'} = $nativeLong eq $bigLong ? 1 : 0;
    return $self -> {'is_big_endian'};
}

sub get_info_fields {
    return (
        'IARL' => 'archivallocation',
        'IART' => 'artist',
        'ICMS' => 'commissioned',
        'ICMT' => 'comments',
        'ICOP' => 'copyright',
        'ICRD' => 'creationdate',
        'IENG' => 'engineers',
        'IGNR' => 'genre',
        'IKEY' => 'keywords',
        'IMED' => 'medium',
        'INAM' => 'name',
        'IPRD' => 'product',
        'ISBJ' => 'subject',
        'ISFT' => 'software',
        'ISRC' => 'supplier',
        'ISRF' => 'source',
        'ITCH' => 'digitizer',
    );
}

sub get_rev_info_fields {
    my $self = shift;
    return %{ $self -> {'rev_info_fields'} } if exists $self -> {'rev_info_fields'};
    my %info_fields = $self -> get_info_fields();
    my %rev_info;
    foreach my $key ( keys %info_fields ) {
        $rev_info{ $info_fields{$key} } = $key;
    }
    $self -> {'rev_info_fields'} = \%rev_info;
    return %rev_info;
}


sub get_sampler_fields {
# dwManufacturer dwProduct dwSamplePeriod dwMIDIUnityNote dwMIDIPitchFraction dwSMPTEFormat dwSMPTEOffset cSampleLoops cbSamplerData
# <sample-loop(s)> <sampler-specific-data> ) <sample-loop> struct dwIdentifier; dwType; dwStart; dwEnd; dwFraction; dwPlayCount;
    return (
        'fields'  => [ qw( manufacturer product sample_period midi_unity_note midi_pitch_fraction smpte_format smpte_offset sample_loops sample_data ) ],
        'loop'    => [ qw( id type start end fraction play_count ) ],
        'extra'   => [],
#        'extra'   => [ map 'unknown' . $_, 1 .. 3  ],
    );
}

sub get_sampler_defaults {
    return (
        'midi_pitch_fraction' => 0,
        'smpte_format'        => 0,
        'smpte_offset'        => 0,
        'product'             => 0,
        'sample_period'       => 0, # 22675,
        'manufacturer'        => 0,
        'sample_data'         => 0,
        'midi_unity_note'     => 65
    );
}

sub get_sampler_loop_defaults {
    return (
        'fraction' => 0,
        'type'     => 0
    );
}


sub error {
    my $self = shift;
    my $filename = shift;
    my $msg = shift;
    my $type = shift;
    my $handler = $self -> {'errorHandler'};
    if ( $handler ) {
        my %hash = (
            'filename' => $filename,
            'message'  => $msg ? $msg : '',
        );
        $hash{'warning'} = 1 if $type eq 'warn';
        &$handler( %hash );
    } else {
        my $txt = $filename ? "$filename: $msg\n" : "$msg\n"; 
        if ( $type && $type eq 'warn' ) {
            warn $txt;
        } else {
            die $txt;
        }
    }
    return;
}

sub get_wav_pack {
    my $self = shift;
    return {
        'order' => [ qw( format channels sample_rate bytes_sec block_align bits_sample ) ],
        'types' => {
            'format'      => 'v',
            'channels'    => 'v',
            'sample_rate' => 'V',
            'bytes_sec'   => 'V',
            'block_align' => 'v',
            'bits_sample' => 'v',
        },
    };
}

1;


Audio-Wav-0.13/Wav/Write.pm
package Audio::Wav::Write;

use strict;
eval { require warnings; }; #it's ok if we can't load warnings

use FileHandle;
use Audio::Wav::Write::Header;

use vars qw( $VERSION );
$VERSION = '0.13';

=head1 NAME

Audio::Wav::Write - Module for writing Microsoft WAV files.

=head1 SYNOPSIS

    use Audio::Wav;

    my $wav = new Audio::Wav;

    my $sample_rate = 44100;
    my $bits_sample = 16;

    my $details = {
	'bits_sample'	=> $bits_sample,
	'sample_rate'	=> $sample_rate,
	'channels'	=> 1,
	# if you'd like this module not to use a write cache, uncomment the next line
	#'no_cache'	=> 1,

    };

    my $write = $wav -> write( 'testout.wav', $details );

    &add_sine( 200, 1 );

    sub add_sine {
	my $hz = shift;
	my $length = shift;
	my $pi = ( 22 / 7 ) * 2;
	$length *= $sample_rate;
	my $max_no =  ( 2 ** $bits_sample ) / 2 - 1;
	for my $pos ( 0 .. $length ) {
	    $time = $pos / $sample_rate;
	    $time *= $hz;
	    my $val = sin $pi * $time;
	    my $samp = $val * $max_no;
	    $write -> write( $samp );
	}
    }

    $write -> finish();

=head1 DESCRIPTION

Currently only writes to a file.

=head1 SEE ALSO

L<Audio::Wav>

L<Audio::Wav::Read>

=head1 NOTES

This module shouldn't be used directly, a blessed object can be returned from L<Audio::Wav>.

=head1 METHODS

=cut

sub new {
    my $class = shift;
    my $out_file = shift;
    my $details = shift;
    my $tools = shift;

    my $handle = (ref $out_file eq 'GLOB') ? $out_file : new FileHandle ">$out_file";

    my $use_cache = 1;
    if ( ref $details eq 'HASH' && exists $details -> {no_cache} ) {
        my $no_cache = delete $details -> {no_cache};
        $use_cache = 0 if $no_cache;
    }

    my $self = {
        'use_cache'   => $use_cache,
        'write_cache' => undef,
        'out_file'    => $out_file,
        'cache_size'  => 4096,
        'handle'      => $handle,
        'details'     => $details,
        'block_align' => $details -> {block_align},
        'tools'       => $tools,
        'done_finish' => 0,
    };

    bless $self, $class;

    unless ( defined $handle ) {
        my $error = $!;
        chomp $error;
        $self -> _error( "unable to open file ($error)" );
        return $self;
    }

    binmode $handle;

    $self -> _init();
    $self -> _start_file();
    $self -> _examine_details( $details );

    if ( $self -> {details} -> {bits_sample} <= 8 ) {
        $self -> {use_offset} = ( 2 ** $self -> {details} -> {bits_sample} ) / 2;
    } else {
        $self -> {use_offset} = 0;
    }

    return $self; 
}

sub DESTROY {
    my $self = shift;
    return unless $self;
    return if $self -> {done_finish};
    $self -> finish();
}

=head2 finish

Finishes off & closes the current wav file.

    $write -> finish();

=cut

sub finish {
    my $self = shift;
    $self -> _purge_cache() if $self -> {use_cache};
    $self -> {header} -> finish( $self -> {pos} );
    $self -> {handle} -> close();
    $self -> {done_finish} = 1;
}

=head2 add_cue

Adds a cue point to the wav file. If $sample is undefined then the position will be the current position (end of all data written so far).

    # $byte_offset for 01 compatibility mode
    $write -> add_cue( $sample, "label", "note"  );

=cut

sub add_cue {
    my $self = shift;
    my $pos = shift;
    my $label = shift;
    my $note = shift;
    my $block_align = $self -> {details} -> {block_align};
    if ( defined $pos ) {
        $pos /= $block_align if $self -> {tools} -> is_01compatible();
    } else {
        $pos = $self -> {pos} / $block_align;
    }
    my $output = {
        'pos' => $pos,
    };
    $output -> {label} = $label if $label;
    $output -> {note} = $note if $note;
    $self -> {header} -> add_cue( $output );
}

=head2 set_sampler_info

All parameters are optional.

    my %info = (
        'midi_pitch_fraction' => 0,
        'smpte_format'        => 0,
        'smpte_offset'        => 0,
        'product'             => 0,
        'sample_period'       => 0,
        'manufacturer'        => 0,
        'sample_data'         => 0,
        'midi_unity_note'     => 65,
    );
    $write -> set_sampler_info( %info );

=cut

sub set_sampler_info {
    my ($self, @args) = @_;
    return $self -> {header} -> set_sampler_info( @args );
}

=head2 add_sampler_loop

All parameters are optional except start & end.

    my $length = $read -> length_samples();
    my( $third, $twothirds ) = map int( $length / $_ ), ( 3, 1.5 );
    my %loop = (
	'start'			=> $third,
	'end'			=> $twothirds,
	'fraction'		=> 0,
	'type'			=> 0,
    );
    $write -> add_sampler_loop( %loop );

=cut

sub add_sampler_loop {
    my ($self, @args) = @_;
    return $self -> {header} -> add_sampler_loop( @args );
}

=head2 add_display

=cut

sub add_display {
    my ($self, @args) = @_;
    return $self -> {header} -> add_display( @args );
}

=head2 set_info

Sets information to be contained in the wav file.

    $write -> set_info( 'artist' => 'Nightmares on Wax', 'name' => 'Mission Venice' );

=cut

sub set_info {
    my ($self, %info) = @_;
    $self -> {details} -> {info} = { %{ $self -> {details} -> {info} }, %info };
}

=head2 file_name

Returns the current filename.

    my $file = $write -> file_name();

=cut

sub file_name {
    my $self = shift;
    return $self -> {out_file};
}

=head2 write

Adds a sample to the current file.

    $write -> write( @sample_channels );

Each element in @sample_channels should be in the range of;

    where $samp_max = ( 2 ** bits_per_sample ) / 2
    -$samp_max to +$samp_max 

=cut

sub write {
    my ($self, @args) = @_;
    my $channels = $self -> {details} -> {channels};
    if ( $self -> {use_offset} ) {
        return $self -> write_raw( pack 'C'.$channels, map { $_ + $self -> {use_offset} } @args );
    } else {
        #TODO: performance: when we move to _init_write_sub, just use:
        #32: pack 'V1', ... 
        #24: substr pack('V1', ...), 3
        #16: pack 'v1', ...
        my $bytes_per_sample = $self->{details}->{bits_sample} >> 3;
        use bytes;
        my @samples = map { substr pack('V1', $_), 0, $bytes_per_sample } @args; 
#warn "bits/sample: $self->{details}->{bits_sample}, bytes/sample: $bytes_per_sample";
#warn "output samples(".scalar @samples."): ".join "-", map ord, split //, join '', @samples;
        return $self -> write_raw( @samples );
    }
}

=head2 write_raw

Adds some pre-packed data to the current file.

    $write -> write_raw( $data, $data_length );

Where;

    $data is the packed data
    $data_length (optional) is the length in bytes of the data

=cut

sub write_raw {
    my $self = shift;
    my $data = shift;
    my $len = shift;
    $len = length $data unless $len;
    return unless $len;
    my $wrote = $len;
    if ( $self -> {use_cache} ) {
        $self -> {write_cache} .= $data;
        my $cache_len = length $self -> {write_cache};
        $self -> _purge_cache( $cache_len ) unless $cache_len < $self -> {cache_size};
    } else {
        $wrote = syswrite $self -> {handle}, $data, $len;
    }

    $self -> {pos} += $wrote;
    return $wrote; 
}

=head2 write_raw_samples

Adds some pre-packed data to the current file, returns number of samples written.

    $write -> write_raw_samples( $data, $data_length );

Where;

    $data is the packed data
    $data_length (optional) is the length in bytes of the data

=cut

sub write_raw_samples {
    my ($self, @args) = @_;
    my $written = $self -> write_raw( @args );
    return $written / $self -> {details} -> {block_align};
}

####################

sub _start_file {
    my $self = shift;
    my( $file, $details, $tools, $handle ) = map { $self -> {$_} } qw( out_file details tools handle );
    my $header = Audio::Wav::Write::Header -> new( $file, $details, $tools, $handle );
    $self -> {header} = $header;
    my $data = $header -> start();
    $self -> write_raw( $data );
    $self -> {pos} = 0;
}

sub _purge_cache {
    my $self = shift;
    my $len = shift;
    return unless $self -> {write_cache};
    my $cache = $self -> {write_cache};
    $len = length $cache unless $len;
    my $res = syswrite $self -> {handle}, $cache, $len;
    $self -> {write_cache} = undef;
}

sub _init {
    my $self = shift;
    my $details = $self -> {details};
    my $output = {};
    my @missing;
    my @needed = qw ( bits_sample channels sample_rate );
    my @wanted = qw ( block_align bytes_sec info wave-ex );

    foreach my $need ( @needed ) {
        if ( exists( $details -> {$need} ) && $details -> {$need} ) {
            $output -> {$need} = $details -> {$need};
        } else {
            push @missing, $need;
        }
    }
    return $self -> _error('I need the following parameters supplied: ' . join ', ', @missing ) if @missing;
    foreach my $want ( @wanted ) {
        next unless ( exists( $details -> {$want} ) && $details -> {$want} );
        $output -> {$want} = $details -> {$want};
    }
    unless ( exists $details -> {block_align} ) {
        my( $channels, $bits ) = map { $output -> {$_} } qw( channels bits_sample );
        my $mod_bits = $bits % 8 ? 1 : 0;
        $mod_bits += int $bits / 8;
        $output -> {block_align} = $channels * $mod_bits;
    }
    unless ( exists $output -> {bytes_sec} ) {
        my( $rate, $block ) = map { $output -> {$_} } qw( sample_rate block_align );
        $output -> {bytes_sec} = $rate * $block;
    }
    unless ( exists $output -> {info} ) {
        $output -> {info} = {};
    }

    $self -> {details} = $output; 
}

sub _examine_details {
    my $self = shift;
    my $details = shift;
    my( $cue, $label, $note ) =
        map { exists( $details -> {$_} ) ? $details -> {$_} : {} }
        qw( cue labl note );
    my $block_align = $self -> {details} -> {block_align};
    my $tools = $self -> {tools};
    foreach my $id ( sort keys %{$cue} ) {       # <-- Thanks to jeremyd713@hotmail.com
        my $pos = $cue -> {$id} -> {position};
        $pos *= $block_align if $tools -> is_01compatible();
        my( $in_label, $in_note ) = 
            map { exists( $_ -> {$id} ) ? $_ -> {$id} : '' }
            ( $label, $note );
        $self -> add_cue( $pos, $in_label, $in_note );
    }
    if ( exists $details -> {sampler} ) {
        my $sampler = $details -> {sampler};
        my $loops = delete $sampler -> {loop};
        $self -> set_sampler_info( %{$sampler} );
        foreach my $loop ( @{$loops} ) {
            $self -> add_sampler_loop( %{$loop} );
        }
    }
    if ( exists $details -> {display} ) {
        my @display = @{ $details -> {display} };
        my @fields = qw( id data );
        $self -> add_display( map { $fields[$_] => $display[$_] } 0, 1 );
    }
}

sub _error {
    my ($self, @args) = @_;
    return $self -> {tools} -> error( $self -> {out_file}, @args );
}

=head1 AUTHORS

    Nick Peskett (see http://www.peskett.co.uk/ for contact details).
    Kurt George Gjerde <kurt.gjerde@media.uib.no>. (0.02-0.03)

=cut

1;


Audio-Wav-0.13/Wav.pm
package Audio::Wav;

use strict;
eval { require warnings; }; #it's ok if we can't load warnings

use Audio::Wav::Tools;

use vars qw( $VERSION );
$VERSION = '0.13';

BEGIN {

    eval { require Inline::C };
    if($@) {
        $Audio::Wav::_has_inline = 0;
    } else {
        # Inline::C is confused with multiple import calls - it seems to 
        # result in errors about @INC. hack around this by launching a
        # seperate process instead of simply checking $@ after:
        #   eval { Inline->import(C => "int foo() { return 0; }\n"); };
        use Config;
        # don't use $^X, which breaks mod_perl - https://rt.cpan.org/Ticket/Display.html?id=62060
        my $path = $Config{perlpath};
        if ($^O ne 'VMS') {
          $path .= $Config{_exe} unless $path =~ m/$Config{_exe}$/i;
        }
        my $inline_c_ok = `$path -e "require Inline::C; eval { Inline->import(C => q[int foo() { return 0; }]) }; print \\\$\@ ? 0 : 1"`;

        if($inline_c_ok) {
            $Audio::Wav::_has_inline = 1;
        } else {
            warn "Inline::C installed, but your C compiler doesn't seem to work with it\n";
            $Audio::Wav::_has_inline = 0;
        }
    }

}

=head1 NAME

Audio::Wav - Modules for reading & writing Microsoft WAV files.

=head1 SYNOPSIS

    # copying a file and adding some cue points to the output file
    use Audio::Wav;
    my $wav = new Audio::Wav;
    my $read = $wav -> read( 'input.wav' );
    my $write = $wav -> write( 'output.wav', $read -> details() );
    print "input is ", $read -> length_seconds(), " seconds long\n";

    $write -> set_info( 'software' => 'Audio::Wav' );
    my $data;
    #read 512 bytes
    while ( defined( $data = $read -> read_raw( 512 ) ) ) {
        $write -> write_raw( $data );
    }
    my $length = $read -> length_samples();
    my( $third, $half, $twothirds ) = map int( $length / $_ ), ( 3, 2, 1.5 );
    my %samp_loop = (
        'start' => $third,
        'end'   => $twothirds,
    );
    $write -> add_sampler_loop( %samp_loop );
    $write -> add_cue( $half, "cue label 1", "cue note 1" );
    $write -> finish();


    # splitting a multi-channel file to separate mono files (slowly!);
    use Audio::Wav;
    my $read = $wav -> read( '4ch.wav' );
    my $details = $read -> details();
    my %out_details = map { $_ => $details -> {$_} } 'bits_sample', 'sample_rate';
    $out_details{channels} = 1;
    my @out_files;
    my $in_channels = $details -> {channels};
    foreach my $channel ( 1 .. $in_channels ) {
        push @out_files, $wav -> write( 'multi_' . $channel . '.wav', \%out_details );
    }

    while ( 1 ) {
        my @channels = $read -> read();
        last unless @channels;
        foreach my $channel_id ( 0 .. $#channels ) {
            $out_files[$channel_id] -> write( $channels[$channel_id] );
        }
    }

    # not entirely necessary as finish is done in DESTROY now (if the file hasn't been finished already).
    foreach my $write ( @out_files ) {
        $write -> finish();
    }


=head1 NOTES

All sample positions are now in sample offsets (unless option '.01compatible' is true).

There is now *very* basic support for WAVEFORMATEXTENSIBLE (in fact it only recognises that the file is in this format).
The key 'wave-ex' is used in the detail hash to denote this format when reading or writing.
I'd like to do more with this, but don't have any hardware or software to test these files, also don't really have any spare time to do the implementation at present.

One day I plan to learn enough C to do the sample reading/ writing in XS, but for the time being it's done using pack/ unpack in Perl and is slow.
Working with the raw format doesn't suffer in this way.

It's likely that reading/ writing files with bit-depth greater than 16 won't work properly, I need to look at this at some point.

=head1 DESCRIPTION

These modules provide a method of reading & writing uncompressed Microsoft WAV files.

=head1 SEE ALSO

    L<Audio::Wav::Read>

    L<Audio::Wav::Write>

=head1 METHODS

=head2 new

Returns a blessed Audio::Wav object.
All the parameters are optional and default to 0

    my %options = (
        '.01compatible'   => 0,
        'oldcooledithack' => 0,
        'debug'           => 0,
    );
    my $wav = Audio::Wav -> new( %options );

=cut

sub new {
    my ($class, @args) = @_;
    my $tools = Audio::Wav::Tools -> new( @args );
    my $self = {
        'tools' => $tools,
    };
    bless $self, $class;
    return $self; 
}

=head2 write

Returns a blessed Audio::Wav::Write object.

    my $details = {
        'bits_sample'	=> 16,
        'sample_rate'	=> 44100,
        'channels'	=> 2,
    };

    my $write = $wav -> write( 'testout.wav', $details );
    my $write = Audio::Wav -> write( 'testout.wav', $details);
    my $write = Audio::Wav -> write( 'testout.wav', $details, %options );

where %options is in the form of arguments for L<Audio::Wav::Tools>.

See L<Audio::Wav::Write> for methods.

=cut

sub write {
    my ($self, $file, $details, @args) = @_;
    require Audio::Wav::Write;
    my $write;
    if(ref $self) {
        $write = Audio::Wav::Write -> new( $file, $details, $self -> {tools} );
    } else {
        $write = Audio::Wav::Write -> new( $file, Audio::Wav::Tools -> new( @args ) );
    }
    return $write; 
}

=head2 read

Returns a blessed Audio::Wav::Read object.

    my $read = $wav -> read( 'testin.wav' );
    my $read = Audio::Wav -> read( 'testin.wav' );
    my $read = Audio::Wav -> read( 'testin.wav', %options );

where %options is in the form of arguments for L<Audio::Wav::Tools>.

See L<Audio::Wav::Read> for methods.

=cut

sub read {
    my ($self, $file, @args) = @_;
    require Audio::Wav::Read;
    my $read;
    if(ref $self) {
        $read = Audio::Wav::Read -> new( $file, $self -> {tools} );
    } else {
        $read = Audio::Wav::Read -> new( $file, Audio::Wav::Tools -> new( @args ) );
    }
    return $read; 
}


=head2 set_error_handler

Specifies a subroutine for catching errors.
The subroutine should take a hash as input. The keys in the hash are 'filename', 'message' (error message), and 'warning'.
If no error handler is set, die and warn will be used.

    sub myErrorHandler {
        my( %parameters ) = @_;
        if ( $parameters{warning} ) {
            # This is a non-critical warning
            warn "Warning: $parameters{filename}: $parameters{message}\n";
        } else {
            # Critical error!
            die "ERROR: $parameters{filename}: $parameters{message}\n";
        }
    }
    $wav -> set_error_handler( \&myErrorHandler );


=cut

sub set_error_handler {
    my ($self, @args) = @_;
    $self -> {tools} -> set_error_handler( @args );
}

=head1 COPYRIGHT

   Copyright (c) 2007-2012 Brian Szymanski <brianski@cpan.org>
    Copyright (c) 1998-2006 Nick Peskett <npeskett@cpan.org> 
    Copyright (c) 2001 Kurt George Gjerde <KJERDE@cpan.org>

=head1 AUTHORS

    Nick Peskett (see http://www.peskett.co.uk/ for contact details).
    Brian Szymanski <ski-cpan@allafrica.com> (0.07-0.13)
    Wolfram humann (pureperl 24 and 32 bit read support in 0.09)
    Kurt George Gjerde <kurt.gjerde@media.uib.no>. (0.02-0.03)

    see also Changes file

=cut

1;
__END__
